### PR TITLE
New shuttle engine features (wip)

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: Shuttle engines now dynamically heat up and cool down. Just being on heats them up, but it'll heat up faster depending on the engine and how powerful it's set. They can be cooled down with water, cold chemicals, and cold gas (Freon is the best at this).
add: Shuttle engines now leave exhaust behind. Also, you'll die if you stand in it.
soundadd: Shuttle engines now have sounds when active.
/:cl:

By the end, all of these will be true:
* Shuttle engines now heat up while they're active.
* Shuttle engines now actually shoot out flames when they're active. The more places for the fire to go, the quicker it cools down on its own.
   * Standing in exhaust does bad things to your body.
* Shuttle speed, launch, etc. can now be controlled from the shuttle console. Although they can't be disabled entirely, you can make the shuttle move slower (safer) or faster (more dangerous, risk of explosion).
   * Central Command regrettably informs you that there is no current plans to install a cooling loop in the shuttle's engines. The engines have been certified to withstand five (5) full minutes of flight time without exploding.*
      * _\*Shuttle safety after arrival at destination not guaranteed._
* You can cool down hot engines with different stuff. The most basic thing is just pumping cold gas into the chamber they're held in, but you can also coat them in cold chemicals. Freon is very good at cooling them and will keep them stable except in the direst situations.

# To-Do List:

- [x] Basics of heat
- [ ] Linked temperature with other engines
- [ ] Cooling systems
- [ ] Engine control from shuttle console?

### CODE REVIEW

**Comment by TehZombehz** [comment]
Can we just rig all shuttles with tripwire bombs every 2 tiles and save us the complexity and time?

**Comment by Xhuis** [comment]
🤔

**Comment by Iamgoofball** [comment]
@TehZombehz don't be a retard

**Comment by scoopscoop** [comment]
Maybe only let'em increase thrust to unsafe levels if the console is emagged. You know how much command staff cares about the crew.

File: code/game/shuttle_engines.dm [hunk 1]
```diff
@@ -1,29 +1,108 @@
...
+			color = rgb(255, 185, 185)
+			playsound(src, 'sound/machines/engine_loop_warm.ogg', 50, 0)
+		if(ENGINE_TEMPERATURE_WARM to ENGINE_TEMPERATURE_MELTING)
+			color = rgb(255, 50, 50)
+			playsound(src, 'sound/machines/engine_loop_melting.ogg', 50, 0)
+		if(ENGINE_TEMPERATURE_EXPLODE to INFINITY)
+			shake_the_room()
+			return
+
+/obj/structure/shuttle/engine/proc/shake_the_room()
```
Comment (optimumtact) [inline code/game/shuttle_engines.dm]: ❤️

**Comment by KorPhaeron** [comment]
>Freon is the best at this

But Freon crashes the server

**Comment by KorPhaeron** [comment]
Will the engines be required for flight? If yes, what happens when they get broken on the escape shuttle? 

Will probably have to remap several shuttle docks as well, for example the mining shuttle docks in the middle of a hallway on efficency station, not a good place for exhaust

**Comment by Xhuis** [comment]
@KorPhaeron To respond to your statements in order from your two comments:

* Freon gas will either be confined to the engine area or just "pumped in" by deleting itself and cooling down the engines directly.

* They will be required to initialize flight. Shuttles will still function if they're destroyed because there is realistically no objects or air to stop the shuttle's movement.

* Exhaust will only shoot out into open space. If there's glass or such in the way it won't pierce. Not realistic but it's for gameplay purposes like the reason you just mentioned.

* A cooling system is in place on the shuttle but it has no gas or other stuff to make it run. If you lug on a canister of cooled oxygen, you can just wrench it in place and be done. As for speed, if you just want to end the round RIGHT NOW you can crank the engines to full power and be there in a minute, or you can slow it down a few if you want more time to do something on the shuttle (i.e. fulfill an objective).

* Shuttles like the cargo one can be sped up to gottagofast speeds to make them take maybe a quarter of the time. The cargo shuttle would arrive and return in thirty seconds. Even though this is much more dangerous, of course, it will save a lot of time for everyone using them. As for rigging the shuttle to blow, yeah, you can do that. It won't stop it but it'll probably kill a few people at the least.

**Comment by Chnkr** [comment]
Please tell me that i can do this

**Comment by Iamgoofball** [comment]
What about shuttles without engines?

**Comment by Supermichael777** [comment]
the drop pod is shot form a railgun

File: code/game/shuttle_engines.dm [hunk 2]
```diff
@@ -1,29 +1,108 @@
...
+			user.emote("scream")
+			return
+		user.visible_message("<span class='notice'>[user] starts unfastening [src]...</span>", "<span class='notice'>You start unfastening [src]...</span>")
+		playsound(src, 'sound/items/Ratchet.ogg', 50, 0)
+		if(!do_after(user, 50, src) || !anchored)
+			return
+		user.visible_message("<span class='notice'>[user] unfastens [src] from the ground!</span>", "<span class='notice'>You undo [src]'s bolts.</span>")
+		anchored = FALSE
+		playsound(src, 'sound/items/Deconstruct.ogg', 50, 0)
+		return
```
Comment (ChangelingRain) [inline code/game/shuttle_engines.dm]: There's a proc for this entire thing, `default_unfasten_wrench`
You can also put the heat check in `can_be_unfasten_wrench`

File: code/game/shuttle_engines.dm [hunk 3]
```diff
@@ -1,29 +1,108 @@
...
+	..()
```
Comment (ChangelingRain) [inline code/game/shuttle_engines.dm]: Also return ..() for attackby

File: code/game/shuttle_engines.dm (hunk 2 above)
Comment (ChangelingRain) [inline code/game/shuttle_engines.dm]: And return 1 here

**Comment by Xhuis** [comment]
@Chnkr Who do you think I am, a casual?
@Iamgoofball No change to those, or maybe I'll add them in.
@Supermichael777 I'm referring to escape pods.

**Comment by Iamgoofball** [comment]
@Xhuis As long as they work without them it's kool

File: code/game/shuttle_engines.dm [hunk 4]
```diff
@@ -1,29 +1,108 @@
...
+			playsound(src, 'sound/machines/engine_loop_normal.ogg', 50, 0)
+		if(ENGINE_TEMPERATURE_STABLE to ENGINE_TEMPERATURE_WARM)
+			color = rgb(255, 185, 185)
+			playsound(src, 'sound/machines/engine_loop_warm.ogg', 50, 0)
+		if(ENGINE_TEMPERATURE_WARM to ENGINE_TEMPERATURE_MELTING)
+			color = rgb(255, 50, 50)
+			playsound(src, 'sound/machines/engine_loop_melting.ogg', 50, 0)
+		if(ENGINE_TEMPERATURE_EXPLODE to INFINITY)
+			shake_the_room()
+			return
```
Comment (RemieRichards) [inline code/game/shuttle_engines.dm]: trailing return, remove it please.

Comment (Iamgoofball) [inline code/game/shuttle_engines.dm]: no

**Comment by Xhuis** [comment]
Sorry, I kind of forgot about this. Expect more stuff later today.

File: code/game/shuttle_engines.dm (hunk 4 above)
Comment (PKPenguin321) [inline code/game/shuttle_engines.dm]: No? More like....


Yes

**Comment by optimumtact** [comment]
WHY CANT YOU OPEN PR'S WHEN THEYRE READY AND NOT BEFORE

**Comment by Iamgoofball** [comment]
@optimumtact get feedback from players

why progress on something if players dont want it etc etc.

**Comment by Xhuis** [comment]
What @Iamgoofball said. What's the point in wasting a ton of effort coding something that nobody wants? I did that with shadowling and I've learned from it.

**Comment by Xhuis** [comment]
Having some issues with shuttle code. I'm not sure how to make all the engines turn on when a shuttle moves, and also for it to detect when to shut off properly. This might take a bit as life is also in the way.

**Comment by optimumtact** [comment]
reeeeeeeeeeeeeeeeee

**Comment by RemieRichards** [comment]
@Xhuis work out the dimensions of the shuttle from it's port, and then loop those dimensions for shuttle engines.

**Comment by Armhulen** [comment]
please stop the "nobody liked shadowling" meme

a lot of people liked it and you know that, even if it *was* favored towards slings

**Comment by KorPhaeron** [comment]
Ping someone when it's ready

**Comment by Xhuis** [comment]
<--- 💩

**Pull request closed without merging**